src/passes/OptimizeInstructions.cpp

```diff
@@ -1804,6 +1823,38 @@ struct OptimizeInstructions
       // Don't bother when `ifFalse` isn't pure - we would need to reverse the
       // order using a temp local, which would be bad
     }
+    {
+      // TODO: Remove this after landing SCCP pass. See: #4161
+
+      // i32(x) ? i32(x) : 0  ==>  x
+      // i32(x) ? 0 : i32(x)  ==>  0
+      Expression *x, *y;
+      if ((matches(curr, select(any(&x), i32(0), any(&y))) ||
+           matches(curr, select(i32(0), pure(&x), any(&y)))) &&
+          areConsecutiveInputsEqualAndFoldable(x, y)) {
```

Comment:
This is not true for the second case. We do not fold them together, we remove x entirely and just leave the 0. Imagine that the first is a tee, and the tee's value is used later.

To handle that, the second case can return `makeSequence(curr->ifFalse, curr->ifTrue)`.

Reply:
Makes sense. I separated this cases and use `areConsecutiveInputsEqualAndFoldable` and `areConsecutiveInputsEqualAndRemovable` for this specific cases